### ISSUE TITLE
Options for only updating user (or item) parameters while keeping the other fixed? 

### ISSUE THREAD

**Issue by dawenl** [lyst/lightfm#300]
This would be useful for strong generalization evaluation. 

I think this should be doable but probably will need to refactor a lot of Cython code to separate user and item parameters update. 

(Trying to rush for a paper submission, otherwise would love to contribute)

**Comment by dawenl** [comment]
I can try that. But wouldn't this also make the gradient itself very large? I am not sure whether `learning_rate * gradient` will end up being 0 or blow up. Any idea?

**Comment by maciejkula** [comment]
No, the gradients are unchanged, so it should end up roughly 0: https://github.com/lyst/lightfm/blob/master/lightfm/_lightfm_fast.pyx.template#L381

**Comment by dawenl** [comment]
In that case, wouldn't setting them to `np.inf` be more precise (i.e., setting the updates to exactly 0)? Or do you see it causing other problems?

**Comment by dawenl** [comment]
It seems this does not work: For strong generalization, you would want to evaluate on a set of users who are not in training, but apparently `fit_partial` was expecting the same set of users from training, while `fit` will overwrite everything. Did I miss something?

**Comment by maciejkula** [comment]
By strong regularization you mean evaluation on a set of users who have been completely excluded from training? This is not straightforward, but is still eminently doable.

You'll need to do the following:
1. Call fit with an interaction matrix that's oversized: say, it has 10,000 rows, but you only have training data for the first 8,000 customers (that is, all the entries in the last 2,000 rows are empty).
2. Freeze the gradients using the method we discussed above.
3. If you're using regularization, you may have to [reinitialize](https://github.com/lyst/lightfm/blob/master/lightfm/lightfm.py#L261) the last 2,000 rows of embeddings to make sure they haven't been regularized to zero.
4. Call `fit_partial` with a 10,000 row matrix: but this time, there will be no data in the first 8,000 rows - only data in the last 2,000 rows.

**Comment by dawenl** [comment]
Thanks! This seems to work (quite a hassle though). Setting to `np.finfo(np.float32).max` seems large enough that the item embeddings didn't change at all.

**Comment by dawenl** [comment]
I think something is off -- I did it one way: call `model.fit_partial` multiple times with `epoch=1`, and after each time, I did all the freezing gradients method we discussed; later I did a `model.fit_partial` once with `epoch=n`, the same `n` which I ran multiple times, and then did the freezing gradients method. The results are drastically different. 

I reported in the paper with the better numbers I got so that I can sleep at night, but would like to solve it in a more principled way.

**Comment by maciejkula** [comment]
Did you freeze before the `epoch=n` run?

**Comment by dawenl** [comment]
For both runs,  the model starts fresh.

**Comment by dawenl** [comment]
Sorry, let me be more clear: 

Here when I call `fit_partial`, I was doing the model training. The only difference is that in the first way, I compute the strong generalization metrics after every epoch by freezing the gradients, while the second way, I only compute the strong generalization metrics after `n` epochs by freezing the gradients.

**Comment by maciejkula** [comment]
OK. So what I think should be happening is this:
1. You run `fit` on your training data.
2. You freeze the model.
3. You run additional `fit_partial` runs to do user fold in.
Is that what you did?

**Comment by maciejkula** [comment]
Once you freeze the gradients the model is shot on the items side unless you unfreeze them, it will not train any more. So you should have gotten _much_ better results after `epochs=n`.

**Comment by dawenl** [comment]
OK, I glossed over the details... I did unfreeze them after each epoch. And interestingly, I got much better results in the first way (`n` calls with `epochs=1`).

**Comment by maciejkula** [comment]
How did you unfreeze them?

**Comment by dawenl** [comment]
I created a copy of the gradients before freezing them and reassign them back after evaluation.

**Comment by dawenl** [comment]
Here is some code snippet: 

```
for _ in range(epochs):
        model.fit_partial(train_data_expanded_coo, epochs=1)
        
        iemb_g_bkp = model.item_embedding_gradients.copy()
        ibias_g_bkp = model.item_bias_gradients.copy()

        model.item_embedding_gradients = np.finfo(np.float32).max * np.ones_like(model.item_embedding_gradients)
        model.item_bias_gradients = np.finfo(np.float32).max * np.ones_like(model.item_bias_gradients)

        model.item_alpha = 0.0
        
        model.user_embeddings[-n_vad_users:] = (
            (model.random_state.rand(n_vad_users, model.no_components) - 0.5) /
            model.no_components).astype(np.float32)
        
        model.fit_partial(vad_data_expanded_coo, epochs=5)
        
        model.item_embedding_gradients = iemb_g_bkp
        model.item_bias_gradients = ibias_g_bkp

        model.item_alpha = alpha
```

**Comment by maciejkula** [comment]
Hmm, in which case the results should be the same. Did you reset regularization as well?

To help more I'd need a reproduction example, I suspect. Maybe try the Movielens dataset and see if the results are similar?

**Comment by maciejkula** [comment]
The code looks solid at first glance.

**Comment by dawenl** [comment]
This is on MovieLens dataset (since i am writing a paper). I will try to get to a reproducible example.

**Comment by maciejkula** [comment]
Awesome! That would be great. I can then dig in.

**Comment by SimonCW** [comment]
I’m closing this issue because it has been inactive for a long time. If you still encounter the problem, please open a new issue.

Thank you!
